rust/backend/ebpf/src/jni_references.rs

```diff
@@ -0,0 +1,146 @@
+    let time_stamp: u64 = unsafe { bpf_ktime_get_ns() };
+    let pid = ctx.pid();
+    let tid = ctx.tgid();
+
+    if unsafe { JNI_REF_PIDS.get(&pid).is_none() } {
+        // don't track calls from this pid
+        return 0;
+    }
+
+    let call = JNICall {
+        pid: pid,
+        tid: tid,
+        begin_time_stamp: time_stamp,
+        method_name: JNIMethodName::AddLocalRef,
+    };
+
+    let mut entry = match JNI_REF_CALLS.reserve::<JNICall>(0) {
+        Some(entry) => entry,
+        None => {
+            error!(&ctx, "could not reserve space in map: JNI_REF_CALLS");
+            return 1;
+        }
+    };
+
+    entry.write(call);
+    entry.submit(0);
+
+    0
+}
+
+#[uprobe]
+pub fn trace_del_local(ctx: ProbeContext) -> u32 {
+    let time_stamp: u64 = unsafe { bpf_ktime_get_ns() };
+    let pid = ctx.pid();
+    let tid = ctx.tgid();
+
+    if unsafe { JNI_REF_PIDS.get(&pid).is_none() } {
+        // don't track calls from this pid
+        return 0;
+    }
+
+    let call = JNICall {
+        pid: pid,
+        tid: tid,
+        begin_time_stamp: time_stamp,
+        method_name: JNIMethodName::DeleteLocalRef,
+    };
+
+    let mut entry = match JNI_REF_CALLS.reserve::<JNICall>(0) {
+        Some(entry) => entry,
+        None => {
+            error!(&ctx, "could not reserve space in map: JNI_REF_CALLS");
+            return 1;
+        }
+    };
+
+    entry.write(call);
+    entry.submit(0);
+
+    0
+}
+
+#[uprobe]
+pub fn trace_add_global(ctx: ProbeContext) -> u32 {
+    let time_stamp: u64 = unsafe { bpf_ktime_get_ns() };
+    let pid = ctx.pid();
+    let tid = ctx.tgid();
+
+    if unsafe { JNI_REF_PIDS.get(&pid).is_none() } {
+        // don't track calls from this pid
+        return 0;
+    }
+
+    let call = JNICall {
+        pid: pid,
+        tid: tid,
+        begin_time_stamp: time_stamp,
+        method_name: JNIMethodName::AddGlobalRef,
+    };
+
+    let mut entry = match JNI_REF_CALLS.reserve::<JNICall>(0) {
+        Some(entry) => entry,
+        None => {
+            error!(&ctx, "could not reserve space in map: JNI_REF_CALLS");
+            return 1;
+        }
+    };
+
+    entry.write(call);
+    entry.submit(0);
+
+    0
+}
+
+#[uprobe]
+pub fn trace_del_global(ctx: ProbeContext) -> u32 {
+    let time_stamp: u64 = unsafe { bpf_ktime_get_ns() };
+    let pid = ctx.pid();
+    let tid = ctx.tgid();
+
+    if unsafe { JNI_REF_PIDS.get(&pid).is_none() } {
+        // don't track calls from this pid
+        return 0;
+    }
+
+    let call = JNICall {
+        pid: pid,
+        tid: tid,
+        begin_time_stamp: time_stamp,
+        method_name: JNIMethodName::DeleteGlobalRef,
+    };
+
+    let mut entry = match JNI_REF_CALLS.reserve::<JNICall>(0) {
+        Some(entry) => entry,
+        None => {
+            error!(&ctx, "could not reserve space in map: JNI_REF_CALLS");
+            return 1;
+        }
+    };
+
+    entry.write(call);
+    entry.submit(0);
+
+    0
+}
```

Comment:
You have 4 ebpf programs that are exactly the same with one difference, the enum variant. 

Two options:

- create a single function that takes the variant and does the logic with 4 entrypoint ebpf programs calling the function with the variant
- create a single ebpf program, retrieve the instruction pointer and map it in userspace

I'd prefer the last option.

Reply:
The variant with the four probes seamed easier for me and as far as I can judge, it is no overhead because the eBPF program gets loaded inline.
I've implemented.